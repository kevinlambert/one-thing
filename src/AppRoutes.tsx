import React from "react";

import PageHome from "./components/page/Home";
import PageThing from "./components/page/Thing";
import Welcome from "./components/page/onboarding/Welcome";
import MomentPage from "./components/page/onboarding/Moment";
import DefaultLayout from "./components/layouts/Default";
import CodeNavigate from "./util/CodeNavigate";

import {
  BrowserRouter,
  Routes,
  Route,
  Outlet,
  useLocation,
} from "react-router-dom";

import { motion, AnimatePresence } from "framer-motion";

const pageTransition = {
  type: "tween",
  ease: "easeInOut",
  duration: 0.4,
};

type Props = {
  children: React.ReactNode;
};

const AnimationLayout = () => {
  const { pathname } = useLocation();
  return (
    <DefaultLayout>
      {/* <AnimatePresence> */}
      <motion.div
        className="motion-wrapper"
        key={pathname}
        initial={{ opacity: 0 }}
        animate={{ opacity: 1 }}
        exit={{ opacity: 0 }}
        transition={pageTransition}
      >
        <Outlet />
      </motion.div>
      {/* </AnimatePresence> */}
    </DefaultLayout>
  );
};

export const routePaths = {
  THING: {
    MOMENT: "/moment",
    VIEW: "/thing/:periodInterval/:periodIncrement",
    EDIT: "/thing/:periodInterval/:periodIncrement/edit",
  },
  HOME: "/",
  ONBOARDING: {
    WELCOME: "/onboarding/welcome",
  },
};

const AppRoutes = () => {
  return (
    <BrowserRouter>
      <Routes>
        <Route element={<CodeNavigate />}>
          <Route element={<AnimationLayout />}>
            <Route path={routePaths.THING.VIEW} element={<PageThing />}></Route>
          </Route>
          <Route
            path={routePaths.THING.EDIT}
            element={<PageThing isEdit={true} />}
          />
          <Route path={routePaths.HOME} element={<PageHome />}></Route>
          <Route
            path={routePaths.ONBOARDING.WELCOME}
            element={<Welcome />}
          ></Route>
          <Route
            path={routePaths.THING.MOMENT}
            element={<MomentPage />}
          ></Route>
        </Route>
      </Routes>
    </BrowserRouter>
  );
};

export default AppRoutes;
