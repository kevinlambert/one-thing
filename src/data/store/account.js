import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import userAPI from "../api/user";

export const fetchAccountByUserID = createAsyncThunk(
  "account/fetchAccountByUserID",
  async (userId) => {
    const data = await userAPI.getUserAccount(userId);
    return data;
  }
);

export const createAccount = createAsyncThunk(
  "account/createAccount",
  async (userId) => {
    const account = await userAPI.createUserAccount(userId);
    const {
      id,
      firstName,
      lastName,
      userID,
      isTourDone,
      termsAndConditionsAccepted,
    } = account;
    return {
      id,
      firstName,
      lastName,
      userID,
      isTourDone,
      termsAndConditionsAccepted,
    };
  }
);

export const updateAccount = createAsyncThunk(
  "account/updateAccount",
  /**  @param arg {any} */
  async ({
    id,
    firstName,
    lastName,
    isTourDone,
    termsAndConditionsAccepted,
  }) => {
    const account = await userAPI.updateUserAccount({
      id,
      firstName,
      lastName,
      isTourDone,
      termsAndConditionsAccepted,
    });
    const { userID } = account;
    return {
      id,
      firstName,
      lastName,
      userID,
      isTourDone,
      termsAndConditionsAccepted,
    };
  }
);

const accountSlice = createSlice({
  name: "account",
  initialState: {},
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchAccountByUserID.fulfilled, (state, action) => {
      return action.payload;
    });

    builder.addCase(createAccount.fulfilled, (state, action) => {
      return action.payload;
    });

    builder.addCase(updateAccount.fulfilled, (state, action) => {
      return action.payload;
    });
  },
});

const { reducer } = accountSlice;

export default reducer;
