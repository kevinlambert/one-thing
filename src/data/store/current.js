import { createSlice } from "@reduxjs/toolkit";

const currentSlice = createSlice({
  name: "current",
  initialState: { currentSphere: null },
  reducers: {
    setCurrentSphere(state, action) {
      state.currentSphere = action.payload;
    },
  },
});

const { actions, reducer } = currentSlice;
export const { setCurrentSphere } = actions;

export default reducer;
