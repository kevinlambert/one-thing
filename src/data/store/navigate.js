import { createSlice } from "@reduxjs/toolkit";

const navigateSlice = createSlice({
  name: "navigate",
  initialState: { navigate: "" },
  reducers: {
    navigateTo(state, action) {
      return action.payload;
    },
  },
});

const { actions, reducer } = navigateSlice;
export const { navigateTo } = actions;
export default reducer;
