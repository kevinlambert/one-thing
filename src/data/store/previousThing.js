import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import thingAPI from "../api/thing";
import { removeNonSerializableValues } from "./util";

export const getPreviousThing = createAsyncThunk(
  "previousThing/getPreviousThing",
  async ({ searchDate, sphereID, periodInterval }) => {
    const data = await thingAPI.getPreviousThing({
      searchDate,
      periodInterval,
      sphereID,
    });

    return removeNonSerializableValues(data);
  }
);

export const updateThing = createAsyncThunk(
  "previousThing/updateThing",
  async ({ id, isDone }) => {
    const data = await thingAPI.updateThing({
      id,
      isDone,
    });

    return removeNonSerializableValues(data);
  }
);

const thingSlice = createSlice({
  name: "previousThing",
  initialState: [],
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(getPreviousThing.fulfilled, (state, action) => {
      if (action.payload.length) {
        const item = action.payload[0]; // should only return 1 item

        const { periodInterval, periodIncrement } = item;
        const { [periodInterval]: statePerdiodInterval, ...restState } = state;

        return {
          [periodInterval]: { [periodIncrement]: item },
          ...restState,
        };
      } else {
        return state;
      }
    });

    builder.addCase(updateThing.fulfilled, (state, action) => {
      const { periodInterval, periodIncrement } = action.payload;
      const { [periodInterval]: statePerdiodInterval, ...restState } = state;

      return {
        [periodInterval]: { [periodIncrement]: action.payload },
        ...restState,
      };
    });
  },
});

const { reducer } = thingSlice;

export default reducer;
