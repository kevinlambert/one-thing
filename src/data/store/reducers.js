import { combineReducers } from "redux";
import userReducer from "./user";
import accountReducer from "./account";
import loadingReducer from "./loading";
import navigateReducer from "./navigate";
import sphereReducer from "./sphere";
import currentReducer from "./current";
import thingReducer from "./thing";
import previousThingReducer from "./previousThing";

export default combineReducers({
  user: userReducer,
  account: accountReducer,
  loading: loadingReducer,
  navigate: navigateReducer,
  spheres: sphereReducer,
  current: currentReducer,
  thing: thingReducer,
  previousThing: previousThingReducer,
});
