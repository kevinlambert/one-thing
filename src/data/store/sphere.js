import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import sphereAPI from "../api/sphere";

export const createSphere = createAsyncThunk(
  "sphere/createSphere",
  async ({ name, isDefault, accountID }) => {
    const data = await sphereAPI.createSphere({
      name,
      isDefault,
      accountID,
    });
    const { id } = data;
    return { id, name, isDefault, accountID };
  }
);

export const getSpheresByAccountID = createAsyncThunk(
  "sphere/getSpheresByAccountID",
  async ({ accountID }) => {
    const data = await sphereAPI.getSpheresByAccountID(accountID);

    return data.map((sphere) => {
      const { id, name, isDefault, accountID } = sphere;
      return { id, name, isDefault, accountID };
    });
  }
);

const sphereSlice = createSlice({
  name: "sphere",
  initialState: [],
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(createSphere.fulfilled, (state, action) => {
      state.push(action.payload);
    });

    builder.addCase(getSpheresByAccountID.fulfilled, (state, action) => {
      return action.payload;
    });
  },
});

const { reducer } = sphereSlice;

export default reducer;
