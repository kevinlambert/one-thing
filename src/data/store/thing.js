import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import thingAPI from "../api/thing";

export const getCurrentThingsBySphere = createAsyncThunk(
  "thing/getCurrentThingsBySphere",
  async ({ currentDate, sphereID, accountID }) => {
    const data = await thingAPI.getCurrentThingsBySphere({
      currentDate,
      sphereID,
      accountID,
    });

    return data;
  }
);

export const saveThing = createAsyncThunk(
  "thing/saveThing",
  async ({
    text,
    periodInterval,
    periodIncrement,
    startDate,
    endDate,
    isDone,
    isRelatedTo,
    sphereID,
    accountID,
  }) => {
    const data = await thingAPI.saveThing({
      text,
      periodInterval,
      periodIncrement,
      startDate,
      endDate,
      isDone,
      isRelatedTo,
      sphereID,
      accountID,
    });

    return data;
  }
);

export const updateThing = createAsyncThunk(
  "thing/updateThing",
  async ({ id, newText, newStartDate, newEndDate, isDone, isRelatedTo }) => {
    const data = await thingAPI.updateThing({
      id,
      newText,
      newStartDate,
      newEndDate,
      isDone,
      isRelatedTo,
    });

    return data;
  }
);

const thingSlice = createSlice({
  name: "thing",
  initialState: [],
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(getCurrentThingsBySphere.fulfilled, (state, action) => {
      return action.payload;
    });

    builder.addCase(saveThing.fulfilled, (state, action) => {
      state.push(action.payload);
    });

    builder.addCase(updateThing.fulfilled, (state, action) => {
      return state.map((item) => {
        return item.id === action.payload.id ? action.payload : item;
      });
    });
  },
});

export const findThingByIntervalIncrementHelper = ({
  things,
  periodInterval,
  periodIncrement,
}) => {
  const result = things.find((item) => {
    return (
      item.periodInterval === periodInterval &&
      item.periodIncrement === periodIncrement
    );
  });

  return result || {};
};

const { reducer } = thingSlice;

export default reducer;
