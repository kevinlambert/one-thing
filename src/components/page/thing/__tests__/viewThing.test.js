import React from "react";
import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import ViewThing from "../ViewThing";
import { Provider as ReduxProvider } from "react-redux";
import * as reactRedux from "react-redux";
import store from "../../../../data/store";
import { routePaths } from "../../../../AppRoutes";

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

test("snapshot", () => {
  const mockStore = {
    thing: [
      {
        id: "6ec8f754-7787-48b1-a934-36819f3307f6",
        text: "doing one thing",
        periodInterval: "day",
        periodIncrement: 0,
        startDate: "2023-10-16",
        endDate: "2023-10-16",
        isRelatedTo: ["WEEK0"],
        isDone: 0,
        sphereID: "8c404182-0b8d-4e8d-ab2d-1c17de42fb7d",
        accountID: "1f2091ec-a047-463b-8251-1a5b2d1f5f63",
        createdAt: "2023-10-16T10:43:59.023Z",
        updatedAt: "2023-10-16T10:46:01.166Z",
        owner: null,
        _version: 4,
        _lastChangedAt: 1697453161214,
        _deleted: null,
      },
    ],
  };

  const useSelectorMock = reactRedux.useSelector;
  useSelectorMock.mockImplementation((selector) => selector(mockStore));

  const div = document.createElement("div");
  const tree = renderer
    .create(
      <ReduxProvider store={store}>
        <MemoryRouter initialEntries={["/thing/day/0"]}>
          <Routes>
            <Route path={routePaths.THING.VIEW} element={<ViewThing />} />
          </Routes>
        </MemoryRouter>
      </ReduxProvider>,
      div
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
