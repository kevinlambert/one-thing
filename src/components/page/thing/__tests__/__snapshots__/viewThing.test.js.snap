// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
Array [
  <div
    className="thing-period-header"
  >
    <div>
      <div
        className="header1"
      >
        Today
      </div>
      <div
        className="subTitle1"
      >
        Monday, Oct 16, 2023
      </div>
    </div>
    <a
      className=""
      href="/thing/day/0/edit"
      onClick={[Function]}
    >
      <div
        className="icon-button icon-button--underline"
      >
        <svg>
          edit_calendar.svg
        </svg>
        <div
          className="icon-button__label"
        >
          Edit
        </div>
      </div>
    </a>
  </div>,
  <div
    className="margin-bottom-16"
  />,
  <div
    className="thing-display"
  >
    doing one thing
  </div>,
  <div
    className="amplify-flex"
    inert={null}
    style={
      Object {
        "justifyContent": "flex-end",
        "marginBlockStart": "var(--amplify-space-medium)",
      }
    }
  >
    <button
      aria-pressed={false}
      className="amplify-button amplify-field-group__control amplify-button--small amplify-togglebutton i-got-this"
      data-fullwidth={false}
      data-size="small"
      inert={null}
      onClick={[Function]}
      style={Object {}}
      type="button"
    >
      I got to this
    </button>
  </div>,
]
`;
